### DISCUSSION TITLE
Natural sort

### DISCUSSION THREAD

**Discussion by spauldhaliwal** [simolus3/drift#2369] [Q&A] [answered]
Currently if I'm ordering say, a list of movies, like so:
`..orderBy([(item) => OrderingTerm(expression: item.title)])`
the output will be something like this:

10 Cloverfield Lane,
127 Hours,
2 Fast 2 Furious,
500 Days of Summer,
7 Prisoners,
Aftersun
Airplane

How can I write a drift expression to order like the following:
2 Fast 2 Furious,
7 Prisoners,
10 Cloverfield Lane,
127 Hours,
500 Days of Summer,
Aftersun,
Airplane

**Comment by FaFre** [comment]
What you describe is a non-standard sort behaviour and I don't know if you really want to go that far apply this to film titles.

One solution would be to extract the number at the beginning of the string with e.g. regex, cast to int and then order by the number first and the original name second. 

Better would be tokenization of the string and then apply a different comparison in case its an int. This is best done in dart and would support mid-string numbers too.

**Comment by simolus3** [comment, marked as answer]
I agree with @FaFre  that this function is probably best written in Dart. There might be an easier way to invoke it through SQL though: You can define a custom collation function for the sqlite3 database with the [createCollation](https://pub.dev/documentation/sqlite3/latest/sqlite3.wasm/CommonDatabase/createCollation.html) method - the [setup](https://drift.simonbinder.eu/api/drift.ffi/databasesetup) callback for `NativeDatabase` would be a good place to do that.

With that setup, you can use ` ..orderBy([(item) => OrderingTerm(expression: item.title.collate(const Collation('yourCollationName')))])` to sort entries according to your custom ordering function.

**Comment by spauldhaliwal** [comment]
Excellent thanks for the suggestions. I'll look into them.

@FaFre you're right that this sorting is non-standard. I assumed it was more widespread, and it actually feels pretty natural to me when I think of alphanumeric sorting. It's actually how Plex sorts its media, however, looking at imdb and tmdb they sort in the more traditional way.

**Comment by spauldhaliwal** [comment]
Actually this was super easy with the Collate method. Collection.dart already has compareNatural, so for posterity's sake:

```dart
NativeDatabase.createInBackground(
      file,
      setup: (database) {
        database.createCollation(
            name: 'naturalTitleSort', function: naturalTitleSort);
      },
    );
```
and
```dart
int naturalTitleSort(String? a, String? b) {
  if (a == null || b == null) {
    return 0;
  } else {
    return compareNatural(a, b);
  }
}
```
finally
```dart
..orderBy([
            (item) => OrderingTerm(
                expression:
                    item.title.collate(const Collate('naturalTitleSort')))
          ])
```

The only issue is I'm now having a problem with my keyset pagination, but I'm sure the fix is simply incorperating the `compareNatural()` function into my `where` clause which I'll take a look at later in the evening.
